Added an option for a latched lid to RoundedBox

rwb27 opened the pull request:
This adds an additional value for `--top` which is `lid_with_latches`. Using this will add an extra plate that sits just inside (i.e. underneath) the lid, with four latches on it. The latches may be moved in and out using four M3 screws.

To assemble, the lid must be glued together with two smaller plates, to make a single unit. One of these plates is the inside of the top of the box (i.e. the part that's normally discarded when `top` is set to `lid` or `open`). That plate has four slots in it, where the screws pass through. This plate is glued directly on to the lid.

The second plate has four "fingers" cut out at the corners. This sits just inside the box, and should be glued onto the plate with the slots. Those two plates are the same size - it's a little tricky to get them lined up precisely, but this is important.

Once the three plates are glued together, the four M3x6mm screws pass through the lid (where there are large clearance slots for the screw heads), through the middle plate (where there are narrower slots that the shaft of the screw can pass through but not the head), and screw into the four fingers. If you slide the screws outwards, the fingers will lock the lid in place.

The holes for the screws need to be threaded before they will work - for my acrylic box I used a tap to cut the thread. Depending on the material it may be possible to self-tap, though that might work better with a different screw. I used M3x6mm screws and they worked neatly. I'm open to feedback on the best kind of screw to specify, though it would of course be easier to test ones I can find in my lab...

I've printed this in 3mm acrylic and it worked fairly well. I can't see why it wouldn't work in plywood/MDF. All the example photos are with wood materials, so I'm guessing if there were to be photos of this, they'd want to be similar?

florianfesti commented:
I like this a lot but I was pondering whether the latches can be done without vitamins. I tried to keep the latches flat so they can also be used on the bottom / with the box upside down. As it does not use round screws in the slots it works without an additional layer for guiding the latches.

I am on vacation this week, so this is still untested.

florianfesti commented:
This needs to be refactored at some point to be usable for other generators with similar lids. But this is out of scope here. Let's get the RoundedBox right first and then think about where else this may be useful.

rwb27 commented:
I've implemented the suggestion above in https://github.com/florianfesti/boxes/pull/827/commits/e0ef44ca3906c9e0273a2490f79d8406e3194bf3. I think it's a fairly minor tweak: I've added hooks to the latches, and increased the length of the slots in the sliding plates to accommodate it. That's moved the edges of those sliding plates slightly closer to the edge of the lid, but I don't think they are problematically close. I've added a screenshot in case that is helpful.

<img height="100" alt="Revised latch shape" src="https://github.com/user-attachments/assets/93b0187d-ce71-4f33-a14c-4a5380547e6b" />

<img height="100" alt="Extended slider" src="https://github.com/user-attachments/assets/ebc1c449-4fee-46e5-803a-02f3b06bcbab" />

I'll try to have a go at cutting this tomorrow, if I can.

rwb27 commented:
I've printed this and it works nicely :) The clicking in and out works really nicely provided the lid is glued in place accurately - and the sliders are really good for helping locate the lid.

I ran into a few issues: first one was that the sliders cut in the wrong order. I needed to use more layers in the laser cutter, so that it first cut out the holes in the sliders, then cut out the sliders, then cut out the lid piece. I don't know if you have a standard way to do that - clearly we've used `@holeCol` already, I think we need a minimum of 3 and ideally 4 layers to get this to work reliably. I realise it doesn't help that I don't have a nice honeycomb bed - but I guess it's relatively simple to output a few different colours.

Also, I glued on the lid after I'd assembled the sliders. That made it much easier to assemble the lid correctly (i.e. get the two plates lined up) but on my first attempt I also glued one of the sliders = game over. I snapped it trying to free the slider. Not quite sure what the solution is to that - probably me being more careful with superglue! I guess a little etching on the lid could create a gutter to stop the glue spreading into the slider, but that may be too much complexity?

Finally, the one drawback of my hook is that the holes in the lid are further out, so you can see through the part when it's taken off the box. I don't think that is a problem, but it's not quite as pretty as without it. Moving everything in by `t` would fix it, but I'm inclined not to unless you think it's important: keeping the latches close to the edge is good for strength.

rwb27 commented:
Oh, and would it be appropriate to add a `--bottom` flag so I can have a latched plate both top and bottom? I would like to do this for the microscope box, though of course I'm subclassing anyway there. I suspect others might find a removable top and bottom useful, though.

florianfesti commented:
--bottom added.
I moved the latches back into the middle. They should still be more than strong enough. I want to use this code for other boxes, too. And some of them might not have as much brim.

florianfesti commented:
Good you brought this up again. I forgot that I wanted to make the slider slightly smaller. That should solve the nesting issue (in theory). Ofc this will depend on the software you are using and whether it will do rounding in a range that is bigger than the current difference.

 Re-cutting would now be very helpful. You can get away with just cutting the lid pieces. The lid should still fit an older box of the same size.

florianfesti commented:
OK, I finally did cut the latches myself. LightBurn at least is able to figure out the right cut order without your patch now that the lines are truly inside of each other.

But the assembly is so messy and risky. I think we need a different solution that doesn't require gluing moving parts in situ. I actually like the latches otherwise. Them not sticking out so they can't be bumped open easily is really nice.
But I guess we need an additional layer somewhere to get the glue away from the sliding surfaces.

I'll try to come up with something.
